File: A2/A2-009/A2-009.cpp
#include<iostream>
#include<vector>
using namespace std;

int N, C;
vector<vector<int>> results;

int main()
{
    // This makes std::cin and std::cout faster
    ios_base::sync_with_stdio(false); cin.tie(0);

    // Input result matrix
    cin >> N >> C;
    results.assign(N, vector<int>(N));
    for(int i = 0; i < N; i++)
    {
        for(int j = 0; j < N; j++)
        {
            cin >> results[i][j];
        }
    }

    // Tournament
    vector<int> thisRound;
    for(int i = 1; i <= N; i++) { thisRound.push_back(i); }
    
    for(int round = N; round > 0; round /= 2)
    {
        vector<int> nextRound;
        for(int i = 0; i < thisRound.size() / 2; i++)
        {
            // Find a winner
            int teamA = thisRound[2 * i];
            int teamB = thisRound[(2 * i) + 1];
            int winner = results[teamA - 1][teamB - 1];

            // Team has the special card
            if(teamA == C && teamA != winner) { winner = teamA; C = 0; }
            if(teamB == C && teamB != winner) { winner = teamB; C = 0; }
            nextRound.push_back(winner);
        }
        // Output
        if(nextRound.size() == 1)
        {
            cout << nextRound[0];
            break;
        }
        // Copy vector
        thisRound = nextRound;
    }
    return 0;
}

File: A2/A2-009/A2-009.py
# Input result matrix
N, C = map(int, input().split())
results = []
for _ in range(N):
    results.append(list(map(int, input().split())))

# Tournament
this_round = []
for i in range(1, N + 1):
    this_round.append(i)

while(N > 0):
    next_round = []
    for i in range(len(this_round) // 2):
        # Find a winner
        teamA = this_round[2 * i]
        teamB = this_round[(2 * i) + 1]
        winner = results[teamA - 1][teamB - 1]
        
        # Team has the special card
        if(teamA == C and teamA != winner):
            winner = teamA
            C = 0
        if(teamB == C and teamB != winner):
            winner = teamB
            C = 0
        next_round.append(winner)
    
    # Output
    if(len(next_round) == 1):
        print(winner)
        break
    this_round = next_round
    N //= 2